### ISSUE TITLE
False positive `no-matching-overload` on classmethod overload

### ISSUE THREAD

**Issue by Apakottur** [astral-sh/ty#3675]
### Summary

Consider the following code:
```python
# a.py

from collections.abc import Sequence
from enum import StrEnum
from typing import Literal, Self, overload


class MyEnum(StrEnum):
    A = "a"


class Model: ...


class Inner[T: MyEnum](Model):
    def kind(self) -> T:
        raise NotImplementedError


class Concrete(Inner[Literal[MyEnum.A]]):
    def kind(self) -> Literal[MyEnum.A]:
        return MyEnum.A


class HasModel[X: Model]:
    @overload
    @classmethod
    def from_model(cls, m: Sequence[X]) -> list[Self]: ...
    @overload
    @classmethod
    def from_model(cls, m: X) -> Self: ...
    @classmethod
    def from_model(cls, m: X | Sequence[X]) -> Self | list[Self]:
        raise NotImplementedError


class Wrapper[T: MyEnum](HasModel[Inner[T]]): ...


def want(x: Wrapper[Literal[MyEnum.A]]) -> None: ...


def repro(c: Concrete) -> None:
    # ty: error[no-matching-overload]; mypy & pyright: ok.
    want(Wrapper.from_model(c))

    # All three accept this:
    want(Wrapper[Literal[MyEnum.A]].from_model(c))
```

mypy and pyright accept it, while ty reports:
```text
a.py:45:10: error[no-matching-overload] No overload of bound method `HasModel.from_model` matches arguments
```

This example is a simplication of a real life use case I have. I suspect it has to do with using `Sequence` in the overload, but am not sure. What am I missing?

### Version

0.0.44

**Comment by Apakottur** [comment]
Slightly smaller example:
```python
# a.py

from typing import Self, overload


class Inner[T]:
    def kind(self) -> T:
        raise NotImplementedError


class Concrete(Inner[int]):
    pass


class HasModel[X]:
    @overload
    @classmethod
    def from_model(cls, m: list[X]) -> Self: ...
    @overload
    @classmethod
    def from_model(cls, m: X) -> Self: ...
    @classmethod
    def from_model(cls, m: X | list[X]) -> Self:
        raise NotImplementedError


class Wrapper[T](HasModel[Inner[T]]): ...


def want(x: Wrapper[int]) -> None: ...


obj = Concrete()

# ty: error[no-matching-overload]; mypy & pyright: ok.
want(Wrapper.from_model(obj))

# All three accept this:
want(Wrapper[int].from_model(obj))
```

**Comment by carljm** [comment]
The root cause here is that we currently don't support specializing a generic class from a call to a classmethod or staticmethod factory. That makes this #541 .
